### PR TITLE
Container interface abstraction

### LINKED ISSUE #3804: Container abstraction planning
Due to recent work on supporting docker swarm for GIEs, it's become apparent that some level of abstraction between container consumers (e.g. GIEs or jobs) and container systems (e.g. docker) would be good, so that we're not writing docker-specific stuff into GIE configs, and so that we can support more complicated environments.

I've also discovered that my initial plan of supporting a swarm that spans both Jetstream zones (TACC and IU) won't work as neatly as planned, because of the reasons outlined in #3802.

The idea would be to build off of the "swarm manager" config I recently added and move those config options to a dict-of-dicts in a more generic config (`containers_conf.yml`, but eventually in `galaxy.yml` when that happens?) e.g.:

```yaml
containers:
  _default_:
    type: docker
  remote:
    type: docker
    command: docker --tlsverify -H tcp://dockerd.example.org:2376/ {docker_args}
```

And a more complex example:

```yaml
containers:
  select_at_random:
    type: random
    group: 
      - jetstream_iu
      - jetstream_tacc
  select_by_function:
    type: python
    python_function: galaxy.containers.rules.select_container
  jetstream_iu:
    type: docker
    docker_command: docker --tlsverify -H tcp://jetstream-iu:2376 {docker_args}
    swarm_mode: yes
  jetstream_tacc:
    type: docker:
    docker_command: docker --tlsverify -H tcp://jetstream-tacc:2376 {docker_args}
    swarm_mode: yes
```

Where `galaxy.containers.rules.select_container()` would take some arguments a la dynamic job rules and return the name of another key in the `containers` dict. 

As for mapping GIEs to containers, we could do this with a dict like:

```yaml
interactive_environment_plugins:
  jupyter
    container: select_by_function
  r_studio:
    container: select_at_random
  phinch:
    container: jetstream_iu
```

It'd be great if this same config worked for jobs running in containers, but jobs are complicated since containers would typically sit at a level beneath a DRM, but might (as in k8) be managed by the DRM.

cc: @erasche @bgruening @jmchilton

### PR DESCRIPTION
**Pull request by natefoo**
Support Docker and Docker Swarm mode (>1.12 SwarmKit). Container interface configuration is performed via the `containers_conf.yml` config file, for which a documented sample is provided.

The GIE launcher has been modified to use the container interface if the `enable_beta_containers_interface` option is set in galaxy.ini. This removes some of the previous GIE-specific swarm mode work that was done.

The Galaxy docker swarm manager also uses the container interface now. It also includes updates for ensuring a minimum number of slots are available for running containers.

The container interface is Python 3 compatible and app-independent.

TODO: The API docker interfaces should be done fairly soon before we have many consumers of the docker model objects, because they'll probably be changed to more closely reflect their API representations. The only reason the CLI versions remain is that they were already written.

Fixes #3804

### CODE REVIEW

File: lib/galaxy/container/docker.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
+"""
+Interface to Docker
+"""
+from __future__ import absolute_import
+
+import logging
+try:
+    from shlex import quote
+except ImportError:
+    from pipes import quote
```
Comment (jmchilton) [inline lib/galaxy/container/docker.py]: Can you use this ``from six.moves import shlex_quote`` instead and in the other place these lines appear?

File: lib/galaxy/container/docker_swarm_manager.py [hunk 2]
```diff
@@ -0,0 +1,560 @@
+"""
```
Comment (jmchilton) [inline lib/galaxy/container/docker_swarm_manager.py]:
```
#!/usr/bin/env python
```

File: lib/galaxy/container/docker_swarm_manager.py [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+    sys.path.insert(0, os.path.abspath(os.path.join(
+        os.path.dirname(__file__),
+        os.pardir,
+        os.pardir)))
+
+from galaxy.config import (
+    configure_logging,
+    find_path,
+    find_root,
+)
```
Comment (jmchilton) [inline lib/galaxy/container/docker_swarm_manager.py]: It seems like there are good abstractions here I'd like to leverage in Pulsar and Planemo. I'd like to just move this all into ``galaxy-lib`` but this script can't be because of these ``galaxy.config`` things.  I could put just a subset of the module in of course - but I usually take this as a sign that a potential library/utility package is attempting to access too much of Galaxy. Maybe it makes some sense to move this one script into ``galaxy.webapps.galaxy`` or keep this script here and move everything else into ``galaxy.util.containers``.

Comment (natefoo) [reply lib/galaxy/container/docker_swarm_manager.py]: I think I'd prefer moving `docker_swarm_manager.py` rather than moving the entire `container` package. And also, we should probably standardize whether packages in galaxy should tend to include a plural form or not.

Comment (natefoo) [reply lib/galaxy/container/docker_swarm_manager.py]: `galaxy.webapps.galaxy` seems like the wrong place for it too, though.

Comment (natefoo) [reply lib/galaxy/container/docker_swarm_manager.py]: Having given this a bit more thought... the only thing the Galaxy stuff is used for is locating the Galaxy config file so that `containers_conf.yml` can be located - except that the automatic loading in `galaxy.container.docker_swarm` doesn't use that option, it specifies the path to `containers_conf.yml` directly via the `--containers-conf-file` option. Outside of that, the swarm manager is useful on its own and apart from Galaxy.

tl;dr I can just remove the Galaxy stuff. Thoughts?

Comment (jmchilton) [reply lib/galaxy/container/docker_swarm_manager.py]: +1 then

**Comment by natefoo** [comment]
@jmchilton Okay, no more Galaxy dependencies in the swarm manager. I also renamed to `galaxy.containers`, squashed and rebased.

**Comment by bgruening** [comment]
This IS awesome! Thanks Nate!!!
@natefoo do you know how to start a IE via API or selenium, so I can test this automatically in Galaxy-Docker?

**Comment by natefoo** [comment]
@bgruening you should only need to hit the path to the IE plugin for an existing dataset, e.g. `/plugins/interactive_environments/jupyter/show?dataset_id=4b187121143038ff`
